### ISSUE TITLE
macOS Sonoma 14.0 x86_64 is failing to compile fortran

### ISSUE THREAD

**Issue by mjsteinbaugh** [iains/gcc-13-branch#9]
Hi @iains ,

I'm hitting a build issue on my Intel MacBook related to fortran that I'm having trouble resolving. The same build script works on my Apple Silicon machine (aarch64 / arm64). And I am using `ld-classic`, as currently recommended. This script worked previosly on macOS Ventura on both architectures.

Build script:
https://github.com/acidgenomics/koopa/blob/main/lang/bash/include/install/common/shared/gcc.sh

```
libtool: compile:  /private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/./gcc/xgcc -B/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/./gcc/ -B/opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/bin/ -B/opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/lib/ -isystem /opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/include -isystem /opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/sys-include -DHAVE_CONFIG_H -I. -I../../../src/libgfortran -iquote../../../src/libgfortran/io -I../../../src/libgfortran/../gcc -I../../../src/libgfortran/../gcc/config -I../../../src/libgfortran/../libquadmath -I../.././gcc -I../../../src/libgfortran/../libgcc -I../libgcc -I../../../src/libgfortran/../libbacktrace -I../libbacktrace -I../libbacktrace -std=gnu11 -Wall -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition -Wextra -Wwrite-strings -Werror=implicit-function-declaration -Werror=vla -fcx-fortran-rules -ffunction-sections -fdata-sections -g -O2 -MT maxval_r10.lo -MD -MP -MF .deps/maxval_r10.Tpo -c ../../../src/libgfortran/generated/maxval_r10.c  -fno-common -DPIC -o .libs/maxval_r10.o
In file included from ../../../src/libgfortran/libgfortran.h:263,
                 from ../../../src/libgfortran/generated/maxval_r10.c:26:
../../../src/libgfortran/generated/maxval_r10.c: In function 'maxval_r10':
./kinds.h:61:26: error: 'l' undeclared (first use in this function)
   61 | #define GFC_REAL_10_HUGE l
      |                          ^
../../../src/libgfortran/generated/maxval_r10.c:154:20: note: in expansion of macro 'GFC_REAL_10_HUGE'
  154 |           *dest = -GFC_REAL_10_HUGE;
      |                    ^~~~~~~~~~~~~~~~
./kinds.h:61:26: note: each undeclared identifier is reported only once for each function it appears in
   61 | #define GFC_REAL_10_HUGE l
      |                          ^
../../../src/libgfortran/generated/maxval_r10.c:154:20: note: in expansion of macro 'GFC_REAL_10_HUGE'
  154 |           *dest = -GFC_REAL_10_HUGE;
      |                    ^~~~~~~~~~~~~~~~
../../../src/libgfortran/generated/maxval_r10.c: In function 'mmaxval_r10':
./kinds.h:61:26: error: 'l' undeclared (first use in this function)
   61 | #define GFC_REAL_10_HUGE l
      |                          ^
../../../src/libgfortran/generated/maxval_r10.c:386:65: note: in expansion of macro 'GFC_REAL_10_HUGE'
  386 |                 result = non_empty_p ? GFC_REAL_10_QUIET_NAN : -GFC_REAL_10_HUGE;
      |                                                                 ^~~~~~~~~~~~~~~~
../../../src/libgfortran/generated/maxval_r10.c: In function 'smaxval_r10':
./kinds.h:61:26: error: 'l' undeclared (first use in this function)
   61 | #define GFC_REAL_10_HUGE l
      |                          ^
../../../src/libgfortran/generated/maxval_r10.c:554:16: note: in expansion of macro 'GFC_REAL_10_HUGE'
  554 |       *dest = -GFC_REAL_10_HUGE;
      |                ^~~~~~~~~~~~~~~~
gmake[3]: *** [Makefile:3678: maxval_r10.lo] Error 1
gmake[3]: *** Waiting for unfinished jobs....
mv -f .deps/maxval_i1.Tpo .deps/maxval_i1.Plo
mv -f .deps/maxval_i2.Tpo .deps/maxval_i2.Plo
mv -f .deps/maxval_i4.Tpo .deps/maxval_i4.Plo
mv -f .deps/maxval_i16.Tpo .deps/maxval_i16.Plo
mv -f .deps/maxval_i8.Tpo .deps/maxval_i8.Plo
libtool: compile:  /private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/./gcc/xgcc -B/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/./gcc/ -B/opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/bin/ -B/opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/lib/ -isystem /opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/include -isystem /opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/sys-include -DHAVE_CONFIG_H -I. -I../../../src/libgfortran -iquote../../../src/libgfortran/io -I../../../src/libgfortran/../gcc -I../../../src/libgfortran/../gcc/config -I../../../src/libgfortran/../libquadmath -I../.././gcc -I../../../src/libgfortran/../libgcc -I../libgcc -I../../../src/libgfortran/../libbacktrace -I../libbacktrace -I../libbacktrace -std=gnu11 -Wall -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition -Wextra -Wwrite-strings -Werror=implicit-function-declaration -Werror=vla -fcx-fortran-rules -ffunction-sections -fdata-sections -g -O2 -MT maxval_r4.lo -MD -MP -MF .deps/maxval_r4.Tpo -c ../../../src/libgfortran/generated/maxval_r4.c -o maxval_r4.o >/dev/null 2>&1
libtool: compile:  /private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/./gcc/xgcc -B/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/./gcc/ -B/opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/bin/ -B/opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/lib/ -isystem /opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/include -isystem /opt/koopa/app/gcc/13.2.0/x86_64-apple-darwin23.0.0/sys-include -DHAVE_CONFIG_H -I. -I../../../src/libgfortran -iquote../../../src/libgfortran/io -I../../../src/libgfortran/../gcc -I../../../src/libgfortran/../gcc/config -I../../../src/libgfortran/../libquadmath -I../.././gcc -I../../../src/libgfortran/../libgcc -I../libgcc -I../../../src/libgfortran/../libbacktrace -I../libbacktrace -I../libbacktrace -std=gnu11 -Wall -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition -Wextra -Wwrite-strings -Werror=implicit-function-declaration -Werror=vla -fcx-fortran-rules -ffunction-sections -fdata-sections -g -O2 -MT maxval_r8.lo -MD -MP -MF .deps/maxval_r8.Tpo -c ../../../src/libgfortran/generated/maxval_r8.c -o maxval_r8.o >/dev/null 2>&1
mv -f .deps/maxval_r8.Tpo .deps/maxval_r8.Plo
mv -f .deps/maxval_r4.Tpo .deps/maxval_r4.Plo
gmake[3]: Leaving directory '/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/x86_64-apple-darwin23.0.0/libgfortran'
gmake[2]: *** [Makefile:1697: all] Error 2
gmake[2]: Leaving directory '/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build/x86_64-apple-darwin23.0.0/libgfortran'
gmake[1]: *** [Makefile:20220: all-target-libgfortran] Error 2
gmake[1]: Leaving directory '/private/var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T/tmp.zDDPtVgZli/build'
gmake: *** [Makefile:1079: all] Error 2
```

I tried looking around online for issues related to `GFC_REAL_10_HUGE` but I'm not coming across any obvious fixes. I'm also looking through the patches in the [Homebrew recipe](https://formulae.brew.sh/formula/gcc), but I'm not seeing anything clearly related to this. Any advice on how to tweak the build?

Best,
Mike

**Comment by iains** [comment]
Which branch are you building, and with what configure line?

I've been building x86-64 trunk/14 on x86_64-Sonoma 14.0 - I guess it's always possible we need to back port some fix (although if it works for Arm64, that might not be the case).

Is this native x86-64 - or via Rosetta2?
(IFF you are using Rosetta, I've found it's necessary to (a) issue an ` arch -x86_64 <shell you use>`  and also make sure that the PATH contains x86_64 tools)

**Comment by mjsteinbaugh** [comment]
Native x86_64

I'm checking if this issue persists with current [gcc-darwin-arm64](https://github.com/iains/gcc-darwin-arm64) code and will post an update.

GCC 13.2 source code: https://github.com/iains/gcc-13-branch/archive/refs/heads/gcc-13-2-darwin.tar.gz

configure args: `-v --disable-nls --enable-checking=release --enable-host-shared --enable-languages=c,c++,fortran,objc,obj-c++ --enable-libstdcxx-time --enable-lto --prefix=/opt/koopa/app/gcc/13.2.0 --with-build-config=bootstrap-debug --with-gcc-major-version-only --with-gmp=/opt/koopa/app/gmp/6.3.0 --with-mpc=/opt/koopa/app/mpc/1.3.1 --with-mpfr=/opt/koopa/app/mpfr/4.2.1 --with-isl=/opt/koopa/app/isl/0.26 --with-zstd=/opt/koopa/app/zstd/1.5.5 --with-boot-ldflags=-static-libstdc++ -static-libgcc  -L/opt/koopa/app/gmp/6.3.0/lib -Wl,-rpath,/opt/koopa/app/gmp/6.3.0/lib -L/opt/koopa/app/mpfr/4.2.1/lib -Wl,-rpath,/opt/koopa/app/mpfr/4.2.1/lib -L/opt/koopa/app/mpc/1.3.1/lib -Wl,-rpath,/opt/koopa/app/mpc/1.3.1/lib -L/opt/koopa/app/isl/0.26/lib -L/opt/koopa/app/gmp/6.3.0/lib -Wl,-rpath,/opt/koopa/app/isl/0.26/lib -L/opt/koopa/app/zstd/1.5.5/lib -Wl,-rpath,/opt/koopa/app/zstd/1.5.5/lib --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk --with-system-zlib --with-ld=/Library/Developer/CommandLineTools/usr/bin/ld-classic`

**Comment by mjsteinbaugh** [comment]
Ah OK, likely related to this change in `math.h`, which affects `x86_64-apple-darwin23.0.0`:
https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;h=93f803d53b5ccaabded9d7b4512b54da81c1c616

Related patch in Homebrew:
https://github.com/Homebrew/homebrew-core/blob/57262c5e2233373537e8b32c00a2b03cec63e7a5/Formula/g/gcc.rb#L19

I'll see if applying this fixes the build. Seems like they had to add that in the Homebrew recipe for Sonoma.

**Comment by iains** [comment]
> Ah OK, likely related to this change in `math.h`, which affects `x86_64-apple-darwin23.0.0`: https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;h=93f803d53b5ccaabded9d7b4512b54da81c1c616
> 
> Related patch in Homebrew: https://github.com/Homebrew/homebrew-core/blob/57262c5e2233373537e8b32c00a2b03cec63e7a5/Formula/g/gcc.rb#L19
> 
> I'll see if applying this fixes the build. Seems like they had to add that in the Homebrew recipe for Sonoma.

Yeah, quite likely - I need to audit which trunk patches need to be back-ported (we are currently nearing the end of stage 1 so that focus is on trunk at present).

**Comment by mjsteinbaugh** [comment]
Argh just tried patching and still seeing the same error. Will do some more sleuthing around online.

**Comment by mjsteinbaugh** [comment]
Unfortunately applying the `math.h` patch alone isn't working for me. Looks like there may be some differences in the Homebrew patch file that differ from the repo here:

https://github.com/Homebrew/formula-patches/blob/master/gcc/gcc-13.2.0.diff

I'll work through it and see if this builds correctly

**Comment by iains** [comment]
> Unfortunately applying the `math.h` patch alone isn't working for me. Looks like there may be some differences in the Homebrew patch file that differ from the repo here:
> 
> https://github.com/Homebrew/formula-patches/blob/master/gcc/gcc-13.2.0.diff
> 
> I'll work through it and see if this builds correctly

I've built a back port stack and bootstrapped it on x86_64 / macOS 12.7 with XC14.3 and x86_64 / macOS 14.0 with XC15b7 / ld-classic .. so it is possible. (testing is in progress).

If that is successful - then I will see about doing a gcc-13.2-release1 .. but maybe it will be at the weekend.

**Comment by mjsteinbaugh** [comment]
Can confirm that I've seeing the same issue when applying the current Homebrew formula patch as well. My CLT version is 15.0.0.0.1.1694021235.

**Comment by iains** [comment]
I cannot repeat your fail with the unpatched 13.2 darwin branch on x86_64 14.0, and my usual configure lines (nor does it fail if I add some of your less contentious configure options)  edit: my CLT is 15-beta-7.

Perhaps there's a mis-configuration;

Golden Rule of configure options "if you do not know exactly what you need something for and exactly why you need it, do not add it", the configure script is intended to make the correct choices for normal operation.    Copying things from distro's scripts is not reliable without researching why the options were added and whether they are still needed. 

> ` --enable-host-shared `  This is not needed for 64b Darwin, it is the default (IFF you are on 32b host, e.g. powerpc or i686 Darwin, which default to mdynamic-no-pic, then you might want to enable it under some circumstances).

>`  --enable-lto `  This is not needed, it is computed by configure based on the host version.
> ` --with-build-config=bootstrap-debug ` not needed, it is the default

> ` --with-boot-ldflags=-static-libstdc++ -static-libgcc -L/opt/koopa/app/gmp/6.3.0/lib -Wl,-rpath,/opt/koopa/app/gmp/6.3.0/lib -L/opt/koopa/app/mpfr/4.2.1/lib -Wl,-rpath,/opt/koopa/app/mpfr/4.2.1/lib -L/opt/koopa/app/mpc/1.3.1/lib -Wl,-rpath,/opt/koopa/app/mpc/1.3.1/lib -L/opt/koopa/app/isl/0.26/lib -L/opt/koopa/app/gmp/6.3.0/lib -Wl,-rpath,/opt/koopa/app/isl/0.26/lib -L/opt/koopa/app/zstd/1.5.5/lib -Wl,-rpath,/opt/koopa/app/zstd/1.5.5/lib`  

^^ AFAICT This is not needed and the first two entries are potentially overriding decisions made by the configure script - why do you think you need all those -rpath entries? - you've given absolute paths to the GMP et. al. installations, so unless you built them in a non-standard way, the libraries already have absolute paths.  I would not be surprised if this option is conflicting/confusing something.

> `--with-native-system-header-dir=/usr/include` What is this meant to be doing? (`/usr/include` is the normal path for native system headers, so AFICT you do not need to specify it).  To use this you need to understand how it interacts with the `--with-sysroot=` option.

I'd suggest simplifying the configure options and re-trying.

**Comment by mjsteinbaugh** [comment]
Thanks @iains for all your help with gcc builds on the Mac. I stripped down my gcc install script to be as simple as possible, still reproducing the same issue above. Builds successfully on Apple Silicon, errors on Intel. I'm guessing this is an issue with Xcode CLT 15.0.0.0.1.1694021235, as I've used this install script for the last couple of macOS releases without issue.

configure args: `-v --enable-languages=c,c++,fortran,objc,obj-c++ --prefix=/opt/koopa/app/gcc/13.2.0 --with-gmp=/opt/koopa/app/gmp/6.3.0 --with-mpc=/opt/koopa/app/mpc/1.3.1 --with-mpfr=/opt/koopa/app/mpfr/4.2.1 --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk --with-system-zlib --with-ld=/Library/Developer/CommandLineTools/usr/bin/ld-classic`

I'm going to close for the time being, and will reopen if this issue persists with the next Xcode CLT release.

Best,
Mike

**Comment by iains** [comment]
FWIW - gcc-13-2 bootstrapped all languages fine with XC CLT 15.1beta - so, hopefully, the 15.1 release will solve your issues.
